### PR TITLE
docs: add info about breaking changes to contributing guide

### LINKED ISSUE #688: What do we define as a breaking change?
This question emerged when @smoya wanted to make the channels field optional - https://github.com/asyncapi/spec/issues/661

This change, might not be a breaking change to the spec itself, however, for tooling that relies on the channels being required and always present, it would.

So when should a change in the spec be considered a breaking change?

I see two ways we can do this (and that is assuming we want to stay with semantic releases).

## Only focusing on the spec
Agreeing that as long as the spec is backward compatible, it is not a breaking change.

## Agreeing what is breaking changes
Agreeing on special cases would probably be possible to minimize breaking changes in tooling and spec.

My immediate thoughts are the following that would be questionable:

_Non-breaking changes:_
- Adding a property

_Breaking changes:_
- Making a property required
- Making a property optional
- Removing a property
- Changing the type of a property in any way

Can you think of another possible solution?

### PR DESCRIPTION
**Pull request by Dule-martins**
I extended the CONTRIBUTING.md file with content for Breaking Change vs Non Breaking Change.
This PR Resolves #688

### CODE REVIEW

File: CONTRIBUTING.md
Comment (fmvilas) [inline CONTRIBUTING.md:83]:
```suggestion
Specifications are not much different than software. Some changes provided in the spec can cause breaking changes for tools that support it. For example, if one of the properties that was `required` becomes `optional`, it is considered a breaking change because some tools might depend on that property and fail if all suddenly it is not provided. 
```

File: CONTRIBUTING.md
Comment (fmvilas) [inline CONTRIBUTING.md:92]: Is it intended to be an exhaustive list or just an example?

**Review by fmvilas — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md
Comment (Dule-martins) [reply CONTRIBUTING.md:92]: from the suggestion made I think it is a list, I can't be sure of how exhaustive it is nor if it is an example.
Maybe @derberg could explain that.

Comment (derberg) [reply CONTRIBUTING.md:92]: it is intended as a result of https://github.com/asyncapi/spec/issues/688

it should be explicit

**Review by fmvilas — DISMISSED** [review]

**Comment by derberg** [comment]
@dalelane @char0n @smoya  as this is a change to contribution guide, I think the rule `to have 3 code owners approval` applies here. 2 more are missing

**Review by smoya — DISMISSED** [review]
LGTM! 🚀🌔

File: CONTRIBUTING.md
Comment (char0n) [inline CONTRIBUTING.md:92]: IMHO this sentence is not precise. I'm suggesting following one:

```suggestion
- Changing the type of a property in backward compatible way
```
**Justification:**: There are multiple ways the field type can change. AsyncAPI 2.x.y is using **UNION** structural typing, thus there are two situations that are similar, but in context of backward compatibility, different.

**Case 1 - backward compatible**

Before Type: `Server Variable Object`
After Type: `Server Variable Object | Reference Object`

**Case 2 - backward incompatible**

Before Type: `Server Variable Object | Reference Object`
After Type: `Server Variable Object`

There is also another case which is variant of `Case 1` and backward compatible as long as `Server Variable Object` and `Difference Object` have the same shape (in context of structural typing). This case can be simplified by calling it a **type renaming**.

**Case 1 (variant 1) - backward compatible**

Before Type: `Server Variable Object`
After Type: `Different Object`

**Review by char0n — CHANGES_REQUESTED** [review]
Proposing a suggestion about type backward compatibility.

File: CONTRIBUTING.md
Comment (fmvilas) [reply CONTRIBUTING.md:92]: Then your suggestion should say "backward incompatible" way, right? 🤔

Comment (char0n) [reply CONTRIBUTING.md:92]: @fmvilas right ;] Thanks, updated the suggestion.

File: CONTRIBUTING.md
Comment (dalelane) [inline CONTRIBUTING.md:86]: should this be adding an optional property?

Comment (fmvilas) [reply CONTRIBUTING.md:86]: Yes, definitely. If you add a required property, it's a breaking change.

Comment (derberg) [reply CONTRIBUTING.md:86]: @dalelane makes sense, can you make a code change suggestion for @Dule-martins please?

Comment (dalelane) [inline CONTRIBUTING.md:86]:
```suggestion
- Adding a new optional property
```

File: CONTRIBUTING.md
Comment (dalelane) [inline CONTRIBUTING.md:88]:
```suggestion
_Breaking changes:_
- Adding a new required property
```

File: CONTRIBUTING.md
Comment (Dule-martins) [reply CONTRIBUTING.md]: @derberg I just committed all suggestions made here.

**Review by smoya — DISMISSED** [review]

**Review by Dule-martins — COMMENTED** [review]
Hi, @fmvilas do you think something here is undone? Can you let me what that is so I could work on it?

File: CONTRIBUTING.md
Comment (dalelane) [inline CONTRIBUTING.md:83]:
```suggestion
Specifications are not much different than software. Some changes provided in the spec can cause breaking changes for tools that support it. For example, if one of the properties that was `required` becomes `optional`, it is considered a breaking change because some tools might depend on that property and fail if it is not provided. 
```

suggested improvement for readability - I don't think it changes the intended meaning

**Review by dalelane — DISMISSED** [review]
I've left a minor suggested wording improvement, but essentially this looks good to me

**Comment by derberg** [comment]
we need to specify something super important about `breaking change` and understanding of `backwards incompatibility`

discussed scenario: 
- I have AsyncAPI document in version 2.4 
- It has  `security` field under operation (not supported in previous versions)
- I switch my version of AsyncAPI document to 2.3

expected result:
- as adding new property is not a breaking change, it is expected that it is a valid AsyncAPI 2.3 document
- in other words, parser written for AsyncAPI 2.3 will still successfully parse such AsyncAPI document with `security` in operation

current result:
- tools fail because JSON Schema for AsyncAPI says `additionalProperties: false`
- specification says explicitly that `additional data can be added to extend the specification` but as `x-` extension

---

This basically means that `new optional filed` is also a breaking changes because it is not backward compatible

more info on 3.0 call -> https://youtu.be/hN6aE3Ebn08?t=1105 (specific moment pointed)

thoughts?

**Comment by GreenRover** [comment]
Recap of today spec 3.0 meeting:
- backward compatibility is not a goal: 
  if the parser is cape able to handle max spec version 2.7 it will not be able to handle 2.8.
  Of the viewpoint tooling every minor spec increase is breaking
- a possible way to define breaking change is:
  you take a spec and change the version number to a newer one and done have to modify something else. The spec is not breaking.

File: CONTRIBUTING.md
Comment (Dule-martins) [reply CONTRIBUTING.md:83]: Yes

**Review by dalelane — APPROVED** [review]

**Review by GreenRover — APPROVED** [review]

**Review by fmvilas — APPROVED** [review]

**Review by derberg — APPROVED** [review]

**Comment by derberg** [comment]
@char0n we are not able to merge without your approval as in the past you requested changes

**Review by char0n — APPROVED** [review]
LGTM

**Comment by char0n** [comment]
@derberg approval provided

**Review by smoya — APPROVED** [review]

**Pull request merged**
